src/Monitor/Tracing/Zipkin.hs

```diff
@@ -140,26 +129,55 @@ new (Settings mbHostname mbPort mbEpt mbMgr mbPrd ignore) = liftIO $ do
       , HTTP.port = maybe 9411 fromIntegral mbPort
       , HTTP.requestHeaders = [("content-type", "application/json")]
       }
-  void $ let prd = fromMaybe 0 mbPrd in if prd <= 0
-    then pure Nothing
-    else fmap Just $ forkIO $ forever $ do
-      threadDelay (microSeconds prd)
-      let flush = flushSpans mbEpt tracer req mgr -- Manager is thread-safe.
-      if ignore
-        then catch flush (\(_ :: HttpException) -> pure ())
-        else flush
-  pure $ Zipkin mgr req tracer mbEpt
+    zpk = Zipkin mgr req tracer mbEpt
+  for_ (microSeconds <$> mfilter (> 0) mbPrd) $ \delay -> fork $ forever $ do
+    threadDelay delay
+    let handler exc = 
+          case fromException exc of
+            (Just (PublishFailed missedSpans _)) -> refillQueue (zipkinTracer zpk) missedSpans
+            Nothing                    -> unless ignore $ throw exc
+    publish zpk `catch` handler
+  pure zpk
+
+refillQueue :: (MonadIO m, MonadBaseControl IO m) => Tracer -> [ZipkinSpan] -> m ()
```

Comment:
For comparison, in [log-elasticsearch](https://github.com/scrive/log/blob/6bf80cab8d9684a6c1dfde3f35edbdff40bd1c40/log-elasticsearch/src/Log/Backend/ElasticSearch.hs#L62) we simply flush the queue, take all messages and try to send them to ES. If this fails with any exception, we just retry in place with the messages we got, there's no need to put these items back into queue unless I'm missing something.

You also then won't need the `PublishFailed` exception IIUC.

Reply:
I'll see how this behaves with an in-place retry.